extension/src/background/helpers/session.ts

```diff
@@ -17,3 +28,249 @@ export class SessionTimer {
+  str: string;
+  keyObject: { iv: ArrayBuffer; key: CryptoKey };
+}
+
+const TEMPORARY_STORE_ENCRYPTION_NAME = "AES-CBC";
+const HASH_KEY_ENCRYPTION_PARAMS = { name: "PBKDF2", hash: "SHA-256" };
+
+export const encryptHashString = ({ str, keyObject }: HashString) => {
+  const encoder = new TextEncoder();
+  const encodedStr = encoder.encode(str);
+
+  return crypto.subtle.encrypt(
+    {
+      name: TEMPORARY_STORE_ENCRYPTION_NAME,
+      iv: keyObject.iv,
+    },
+    keyObject.key,
+    encodedStr,
+  );
+};
+
+interface DecodeHashString {
+  hash: ArrayBuffer;
+  keyObject: { iv: ArrayBuffer; key: CryptoKey };
+}
+
+export const decryptHashString = async ({
+  hash,
+  keyObject,
+}: DecodeHashString) => {
+  const decrypted = await crypto.subtle.decrypt(
+    {
+      name: TEMPORARY_STORE_ENCRYPTION_NAME,
+      iv: keyObject.iv,
+    },
+    keyObject.key,
+    hash,
+  );
+
+  const textDecoder = new TextDecoder();
+
+  return textDecoder.decode(decrypted);
+};
+
+export const deriveKeyFromString = async (str: string) => {
+  const iterations = 1000;
+  const keylen = 32;
+  const keyLength = 48;
+  // randomized salt will make sure the hashed password is different on every login
+  const salt = crypto.getRandomValues(new Uint8Array(16)).toString();
+
+  const encoder = new TextEncoder();
+  const keyMaterial = encoder.encode(str);
+
+  const importedKey = await crypto.subtle.importKey(
+    "raw",
+    keyMaterial,
+    HASH_KEY_ENCRYPTION_PARAMS,
+    false,
+    ["deriveBits"],
+  );
+
+  const saltBuffer = encoder.encode(salt);
+  const params = {
+    ...HASH_KEY_ENCRYPTION_PARAMS,
+    salt: saltBuffer,
+    iterations,
+  };
+  const derivation = await crypto.subtle.deriveBits(
+    params,
+    importedKey,
+    keyLength * 8,
+  );
+
+  const derivedKey = derivation.slice(0, keylen);
+  const iv = derivation.slice(keylen);
+
+  const importedEncryptionKey = await crypto.subtle.importKey(
+    "raw",
+    derivedKey,
+    { name: TEMPORARY_STORE_ENCRYPTION_NAME },
+    true,
+    ["encrypt", "decrypt"],
+  );
+
+  return {
+    key: importedEncryptionKey,
+    iv,
+  };
+};
+
+interface StoreActiveHashKey {
+  sessionStore: Store;
+  hashKey: {
+    key: CryptoKey;
+    iv: ArrayBuffer;
+  };
+}
+
+export const storeActiveHashKey = async ({
+  sessionStore,
+  hashKey,
+}: StoreActiveHashKey) => {
+  const format = "jwk"; // JSON Web Key format
+  // export the key for transferability
+  const exportedKey = await crypto.subtle.exportKey(format, hashKey.key);
+
+  // store hashed password in memory
+  sessionStore.dispatch(
+    setActiveHashKey({
+      hashKey: {
+        // properly encode ArrayBuffer into serializable format
+        iv: encode(hashKey.iv),
+        // JSON Web Key is able to be stringified without encoding
+        key: JSON.stringify(exportedKey),
+      },
+    }),
+  );
+};
+
+interface GetActiveHashKey {
+  sessionStore: Store;
+}
+
+export const getActiveHashKeyCryptoKey = async ({
```

Comment:
this does the reverse of above - get the encode version of the CryptoKey from redux and then decode it into something we can use for encryption/decryption